Fix unboxing stubs on wasm

radekdoulik opened the pull request:
Fix https://github.com/dotnet/runtime/issues/120092

radekdoulik committed:
Fix unboxing stubs on wasm

radekdoulik commented:
This allows following code to run
```
TestCallingConvention0(1, 3, 5);

static void TestCallingConvention0(int a, int c, int e)
{
    Console.WriteLine("TestCallingConvention0: a = {0}, c = {1}, e = {2}", a, c, e);
}
```

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1727,13 +1729,16 @@ Stub * MakeUnboxingStubWorker(MethodDesc *pMD)
 #endif
     {
+#ifdef FEATURE_PORTABLE_ENTRYPOINTS
+            pstub = CreateUnboxingILStubForValueTypeMethods(pUnboxedMD);
```
```suggestion
        pstub = CreateUnboxingILStubForValueTypeMethods(pUnboxedMD);
```

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2261,6 +2267,16 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT, CallerGCMode callerGCMo
     {
         pStub = MakeUnboxingStubWorker(this);
+
+#if defined(FEATURE_PORTABLE_ENTRYPOINTS) 
```
This should be moved down and replace `if (!GetOrCreatePrecode()->SetTargetInterlocked(pStub->GetEntryPoint()))` so that it applies to all paths the return `Stub*`.

jkotas commented on src/coreclr/vm/prestub.cpp:
You should be able to delete `EmitComputedInstantiatingMethodStub` and `EmitShuffleThunk` from wasm\cgencpu.h

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2261,6 +2267,16 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT, CallerGCMode callerGCMo
+
+#if defined(FEATURE_PORTABLE_ENTRYPOINTS) 
+        pCode = pStub->GetEntryPoint();
+        delete pStub;
+        pStub = NULL;
```
`Stub` is just a temporary wrapper for the IL-based stubs. It is fine to delete it. It is not registered anywhere.

However, it should be done later in the method, so that the code applies to all paths that use Stub (see my other comment).

radekdoulik replied:
I have added call to `pStub->DecRef();` to not leak the allocations inside the Stub.

radekdoulik replied:
Hmm, I should call just that and not delete the Stub.

radekdoulik committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>
Co-authored-by: Aaron Robinson <arobins@microsoft.com>

radekdoulik committed:
Feedback

Remove GenerateShuffleArray and s_pShuffleThunkCache for FEATURE_PORTABLE_ENTRYPOINTS

radekdoulik committed:
Feedback

Handle all pStubs for FEATURE_PORTABLE_ENTRYPOINTS

radekdoulik committed:
Do not double free

radekdoulik committed:
Add thunk for the IL stub

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2369,12 +2376,27 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT, CallerGCMode callerGCMo
+#ifndef FEATURE_PORTABLE_ENTRYPOINTS
     else
     {
         if (!GetOrCreatePrecode()->SetTargetInterlocked(pStub->GetEntryPoint()))
```
```suggestion
else
{
#ifdef FEATURE_PORTABLE_ENTRYPOINTS
        pCode = pStub->GetEntryPoint();
        pStub->DecRef();

        void* ilStubInterpData = PortableEntryPoint::GetInterpreterData(pCode);
        _ASSERTE(ilStubInterpData != NULL);
        SetInterpreterCode((InterpByteCodeStart*)ilStubInterpData);
#else
        if (!GetOrCreatePrecode()->SetTargetInterlocked(pStub->GetEntryPoint()))
```
Move this here to have just have one ifdef?

radekdoulik replied:
I guess we can do this if we add call to `SetCodeEntryPoint(pCode);`

radekdoulik commented:
I also added new thunk for the IL stub. I wonder if we can avoid that later. By replacing calli with call inside the stub?

jkotas commented on src/coreclr/vm/wasm/helpers.cpp:
```diff
@@ -553,6 +553,13 @@ namespace
 
+    void CallFunc_I32_I32IND_I32_I32IND_I32_RetI32(PCODE pcode, int8_t *pArgs, int8_t *pRet)
+    {
+        printf("TRACE CallFunc_I32_I32IND_I32_I32IND_I32_RetI32 pcode: %lx arg0: %x arg1: %x arg2: %x\n", pcode, ARG(0), ARG_IND(1), ARG_IND(2));
```
```suggestion
```
Left-over?

radekdoulik committed:
Update src/coreclr/vm/wasm/helpers.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

radekdoulik committed:
Update src/coreclr/vm/prestub.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

radekdoulik committed:
Finish the feedback

radekdoulik committed:
Feedback

jkotas commented on src/coreclr/vm/comdelegate.h:
```diff
@@ -186,7 +186,11 @@ class ShuffleThunkCache : public StubCacheBase
     {
         STANDARD_VM_CONTRACT;
 
+#ifndef FEATURE_PORTABLE_ENTRYPOINTS
```
Can this ifdef be around the whole ShuffleThunkCache type?

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2377,6 +2384,15 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT, CallerGCMode callerGCMo
+        void* ilStubInterpData = PortableEntryPoint::GetInterpreterData(pCode);
+        _ASSERTE(ilStubInterpData != NULL);
+        SetInterpreterCode((InterpByteCodeStart*)ilStubInterpData);
+        SetCodeEntryPoint(pCode);
```
Are both `SetInterpreterCode` and `SetCodeEntryPoint` needed? I would expect that you need only one.

radekdoulik replied:
When I don't set interpreter code I am getting
```
ASSERT FAILED
        Expression: portableEntryPoint->HasNativeCode()
        Location:   line 28 in /Users/rodo/git/runtime-main/src/coreclr/vm/precode_portable.cpp
        Function:   GetActualCode
        Process:    42
Frame (InterpreterFrame): 0x4fee58
   0) System.Text.ValueStringBuilder::AppendFormatHelper, IR_051c
   1) System.IO.StreamWriter::WriteFormatHelper, IR_007c
   2) System.IO.StreamWriter::WriteLine, IR_004d
   3) System.Console::WriteLine, IR_0018
   4) Program::<<Main>$>g__TestCallingConvention0|0_0, IR_0023
   5) Program::<Main>$, IR_0011
```
and when I don't set code entry point, it triggers the 2 asserts at the end of this method.

jkotas replied:
Ok

radekdoulik committed:
Feedback

radekdoulik committed:
Feedback

radekdoulik commented on src/coreclr/inc/switches.h:
```diff
@@ -165,6 +165,10 @@
+#if !defined(FEATURE_PORTABLE_ENTRYPOINTS) && !defined(TARGET_X86)
+#define FEATURE_PORTABLE_SHUFFLE_THUNKS
+#endif
+
```
Why not `#if defined(TARGET_UNIX) || !defined(TARGET_X86)`?

radekdoulik committed:
Feedback

jkotas commented:
/ba-g missing logs, opened #120174 and #120176 on failures with logs
